### ISSUE TITLE
Cannot use "--amend-rooms" - results in HTTP 404

### ISSUE THREAD

**Issue by anx-ag** [ergon/migratemost#12]
Hey there,

when I try to use --amend-rooms I do get an HTTP 404 and am not sure about the reason. When I do not use this parameter, everything else works so far.

Any ideas?

Here's the output:

migratemost/migratemost.py  \
-t TEAMNAME  \
-o /data/import  \
-i /data/hipchat/overall  \
-v  \
--migrate-all  \
--amend-rooms \
--public-channel-membership-based-on-redis-export  \
--map-town-square-channel="blahblah"  \
--authentication-service=ldap  \
--authentication-data-field=username  \
--concat-output  \
--migrate-custom-emoticons  \
--hipchat-base-url=https://hipchat.mydomain.com/v2/  \
--hipchat-access-tokens=top,secret,list,of,tokens  \
--disable-tutorial  \
--shrink-image-to-limit

2019-09-17 14:29:15,392 - INFO - Starting migration
2019-09-17 14:29:15,392 - INFO - Amending Hipchat room export
2019-09-17 14:29:15,392 - INFO - Starting to amend Hipchat room export (this might take a while)
2019-09-17 14:29:39,421 - ERROR - HTTP Error 404: Not Found

Thanks!

**Comment by 7-plus-t** [comment]
Hey @anx-ag 

In general, it looks like a REST call to Hipchat failed.

Just to verify: Does running the script without the `--amend-rooms` option work, including exporting the emoticons from Hipchat? If so, that would let me assume that the Hipchat URL and tokens are working. 

You could try to run the `amend_hipchat_rooms.py` script directly and see what the output of that is. It should give you more details.

BTW: A different option is to export the room memberships from Redis directly instead of fetching them from the Hipchat API. In my experience that result is a bit more accurate.

**Comment by anx-ag** [comment]
Hey,

thanks for the fast reply - I'm already using the redis option (as you can see in the commands and I've also created the autojoin.json file). Does that mean that I can omit --amend-rooms when I use the redis export?

And yes, omitting --amend-rooms works flawlessly (except for some other issues which I will summarize here once I managed to complete the migration).

**Comment by 7-plus-t** [comment]
Hi

Sorry, I missed the `--public-channel-membership-based-on-redis-export` in your first comment. **TL;DR** Yes, you can skip `--amend-rooms` when using `--public-channel-membership-based-on-redis-export`. 

Basically both options try to improve the accuracy of the members and participant lists in the Hipchat export. In my experience, members is accurate in the export but participants are missing. Your mileage may vary. ('members': users with membership in a room, relevant for private rooms only; 'participants': users currently in a room)

**Comment by 7-plus-t** [comment]
>  (except for some other issues which I will summarize here once I managed to complete the migration)

Looking forward to your feedback 👍

**Comment by 7-plus-t** [comment]
btw: you can find my on the [Mattermost community server](https://pre-release.mattermost.com) as seven-plus-t

**Comment by anx-ag** [comment]
Alright, great! Then I'll leave out the --amend-rooms option for now in the hope that group/channel memberships don't look that wrong at the end :)

I did try the "amend_hipchat_rooms.py" script now and it throws the same error, here's the output, JFYI:

 ./amend_hipchat_rooms.py -b https://hipchat.mydomain.com/v2/ -i /data/hipchat/overall/rooms.json -o /tmp -t topsecret
2019-09-17 17:06:05,439 - INFO - Starting to amend Hipchat room export (this might take a while)
2019-09-17 17:21:26,583 - ERROR - HTTP Error 404: Not Found

**Comment by 7-plus-t** [comment]
To get more detailed output, you need to set the logging level in the script to `logger.setLevel(logging.DEBUG)` (no parameter to set it, sorry).

**Comment by anx-ag** [comment]
Here's the debug output when running the pythong script with the DEBUG loglevel enabled:

2019-09-17 19:00:21,899 - DEBUG - updating room: 279: xyz-721
2019-09-17 19:00:21,953 - DEBUG - updating room: 280: Test
2019-09-17 19:00:21,976 - ERROR - HTTP Error 404: Not Found

So it seems as if it is not able to find a room. Here's the JSON (out of rooms.json) for this specific room:

```
  {
    "Room": {
      "created": "2019-01-23T10:09:40+00:00",
      "guest_access_url": null,
      "id": 280,
      "is_archived": false,
      "members": [
        36
      ],
      "name": "Test",
      "owner": 36,
      "participants": [],
      "privacy": "private",
      "room_admins": [
        36
      ],
      "topic": ""
    }
  },

```
I do have an exported history for this room (/data/hipchat/export/rooms/280/history.json), so it must at least exist somehow.

**Comment by anx-ag** [comment]
> **TL;DR** Yes, you can skip `--amend-rooms` when using `--public-channel-membership-based-on-redis-export`.

vs.

https://github.com/ergon/migratemost/blob/master/HOWTO.md#example

BTW :)

**Comment by 7-plus-t** [comment]
Fixed the example, thanks for the hint :)

The room looks perfectly fine, I guess you would have to check the log on your Hipchat instance to see why it returned a 404. Maybe it got deleted/archived in the meantime? The Hipchat API does not `/member` call on rooms that are not active.

**Comment by anx-ag** [comment]
What URLs is migratemost trying to request? I can then check them manually here to see where it starts to fail.
We do have tons of at least archived rooms, not sure about deleted ones though.

**Comment by 7-plus-t** [comment]
Assuming the first of two calls fail, it is this one: https://hipchat.mydomain.com/v2/room/<room_id>/member?max-results=1000&auth_token=<secret_token>
You could also check the state of the room in the Hipchat admin web UI and see if it is archived/deleted or something else looks suspicious with it.

The script skips rooms that are marked as archived in the export, so unless the exported data is outdated or Hipchat didn't set the flag correctly, it should not do the request.

**Comment by anx-ag** [comment]
Room ID 280:

```
error |  
-- | --
code | 404
message | "Room '280' not found"
type | "Not Found"
```

An archived room looks like this:

```
error |  
-- | --
code | 400
message | "Cannot access an archived room"
type | "Bad Request"
```

So whatever that is, but it seems to be a bug in the hipchat database, also in the hipchat admin UI it says "room not found", so I guess the best idea would be to ignore resources with http error code 404 (at least for the rooms).

So I did patch hipchat_api.py and amend_hipchat_rooms.py now to ignore 400 and 404 errors on fetch_member calls for the rooms, because I experienced both situations in my database: 404 for a room which is still mentioned in the database (the reason for this ticket) and a 400 for a room which was incorrectly flagged as active, while it is archived.


```
--- hipchat_api.py.orig 2019-09-20 22:08:56.927523443 +0200
+++ hipchat_api.py      2019-09-20 22:10:50.606895025 +0200
@@ -49,7 +49,7 @@
         else:
             raise e

-def fetch_and_parse(url, access_tokens):
+def fetch_and_parse(url, access_tokens, ignorecodes = []):
     global tokens
     global current_token_index
     tokens = access_tokens
@@ -60,5 +60,8 @@
         body = response.read()
         return json.loads(body)
     except urllib2.URLError as e:
-        logger.error(e)
-        exit(1)
+        if e.code in ignorecodes:
+            return None
+        else:
+            logger.error(e)
+            exit(1)
```

```
--- amend_hipchat_rooms.py.orig 2019-09-20 22:10:55.923051911 +0200
+++ amend_hipchat_rooms.py      2019-09-20 22:53:09.685441625 +0200
@@ -20,8 +20,11 @@
 option_output_path = './'

 def _fetch_members(base_url, tokens, room_id):
-    members = hipchat_api.fetch_and_parse(base_url + 'room/%d/member?max-results=1000' % room_id, tokens)
-    return list(map(lambda m: m[u'id'], members[u'items']))
+    members = hipchat_api.fetch_and_parse(base_url + 'room/%d/member?max-results=1000' % room_id, tokens, [ 400, 404 ])
+    if members is None:
+        return None
+    else:
+        return list(map(lambda m: m[u'id'], members[u'items']))

 def _fetch_participants(base_url, tokens, room_id):
     participants = hipchat_api.fetch_and_parse(base_url + 'room/%d/participant?max-results=1000' % room_id, tokens)
@@ -101,17 +104,22 @@

     logger.info('Starting to amend Hipchat room export (this might take a while)')
     rooms = _load_hipchat_rooms(input_file)
-    for room_container in rooms:
-        r = room_container[u'Room']
+    for i in xrange(len(rooms) - 1, -1, -1):
+        r = rooms[i][u'Room']
         room_id = int(r[u'id'])
         logger.debug('updating room: %d: %s' % (room_id, r[u'name']))
-
         if r[u'is_archived']:
             # skip archived rooms. API does not allow to fetch members for archived rooms.
             logger.debug('\tskipping archived room')
             continue

-        r[u'members'] = _fetch_members(hipchat_base_url, hipchat_tokens, room_id)
+        _hc_temp_members = _fetch_members(hipchat_base_url, hipchat_tokens, room_id)
+        if _hc_temp_members is None:
+            logger.debug('\tskipping deleted/corrupt room')
+            del rooms[i]
+            continue
+
+        r[u'members'] = _hc_temp_members
         r['participants'] = _fetch_participants(hipchat_base_url, hipchat_tokens, room_id)

```

**Comment by anx-ag** [comment]
But the question still is: I do now have the hc_amend_rooms.json file - what should I do with it now? :) Can I use it to compare the room memberships between redis and the --amend-rooms option now somehow? The reason I'm asking is that I recognized that all the previous guest accounts (which are deactivated now) are still members of the rooms, so it seems as if inactive accounts are not removed from the rooms after the import or should they be and this is another special case on my installation or did I miss a parameter somewhere?

Thanks!

P.S.: will provide a new patch, just recognized that the temp variable is not necessary anymore, i had to use it when I was first trying to work directly in the list, which was not of much success.

**Comment by 7-plus-t** [comment]
Hi again :)

First of all, thanks for the patch! Will integrate it as soon as I get the chance. 
We did not use guest accounts, so I had no data to implement their handling in Migratemost. In general, Hipchat has a "funny" handling of room memberships and this sounds like one more quirk in it. 

Migratemost uses the participant/membership files like this:
1. Take rooms from Hipchat export
2. If option `--amend_rooms` was used, call `amend_hipchat_rooms.py` and overwrite `members` and `participants` of the rooms with the data fetched from the running Hipchat installation
3. If option `--public-channel-membership-based-on-redis-export` was used, add `participants` found in the Redis file

So you could look at the Redis file and see if it has the more accurate participants lists than the running Hipchat server/the Hipchat export and if so, use both options.

Concerning the guest accounts: If you import the guest accounts to Mattermost and then deactivate them afterwards, I would expect Mattermost to remove them from the channels. Could that work? 

PS: When I implemented Migratemost, there was no option to mark users as deleted during bulk import, therefore the post-import manual step to deactivate them.

**Comment by anx-ag** [comment]
I think there's still no way to mark them as deleted? Anyways, you were right, once the users are deactivated, they do automatically get kicked out of the channels, so this is working as expected. I just do not know why some of the guest accounts have been inactive right from the beginning after the import and some had to be deactivated manually, need to figure that out, but besides that, it looks good so far.

**Comment by 7-plus-t** [comment]
> I think there's still no way to mark them as deleted?

According to https://docs.mattermost.com/deployment/bulk-loading.html#user-object there is a field `delete_at` to set the timestamp for when a user was deactivated.

> I just do not know why some of the guest accounts have been inactive right from the beginning after the import and some had to be deactivated manually, need to figure that out, but besides that, it looks good so far.

Did you figure out what happened there?

**Comment by 7-plus-t** [comment]
Closing due to inactivity.
